DirectProgramming/DPC++FPGA/ReferenceDesigns/cholesky_inversion/src/cholesky_inversion_demo.cpp

```diff
@@ -0,0 +1,447 @@
+    std::vector<T> i_matrix;
+    std::vector<T> r, weights;
+
+    a_matrix.resize(kAMatrixSize * kMatricesToInvert);
+    i_matrix.resize(kIMatrixSize * kMatricesToInvert);
+
+    std::cout << "Generating " << kMatricesToInvert << " random ";
+    if constexpr (kComplex) {
+      std::cout << "complex ";
+    } else {
+      std::cout << "real ";
+    }
+    std::cout << "matri" << (kMatricesToInvert > 1 ? "ces" : "x") << " of size "
+              << kRows << "x" << kColumns << " " << std::endl;
+
+    // Generate the random (Hermitian and positive-definite) input matrices
+    srand(kRandomSeed);
+
+    // Max condition number
+    constexpr float kEpsilon = 0.5;
+
+    // Random min and max values for the random floating-point value
+    // generation
+    constexpr float kRandomMin = 0;
+    constexpr float kRandomMax = 1;
+
+    /*
+      Generate a random matrix with a given epsilon such that
+      cond(M, inf) <= (1+epsilon)/(1-epsilon)
+      This is helpful as having a condition number with infinite norm close to 1
+      improves the numerical stability of the matrix inversion.
+      Provided an epsilon value, this function populates the output vector with
+      a matrix in a row fashion.
+
+      Algorithm courtesy of Carl Christian Kjelgaard Mikkelsen (spock@cs.umu.se)
+
+      Matlab code snippet this function reimplements in C++:
+
+        function [A, B]=myDiagonal(m,epsilon)
+
+        % Returns a matrix that is diagonally dominant by rows
+        %
+        % CALL SEQUENCE:
+        %    [A, B]=ccDiagonally(m, epsilon)
+        %
+        % INPUT:
+        %    m        the dimension
+        %    epsilon  the dominance factor epsilon in (0,1]
+        %
+        % OUTPUT:
+        %    A        a matrix which is strictly diagonally dominant by rows
+        %    B        B = D\A, where D is the diagonal of A
+        %
+        % The main purpose of this function is to construct test matrices
+        % for, say, Gaussian elimination with no pivoting.
+        %
+        % The matrix A is not necessarily well-conditioned, but
+        % the infinity norm condition number of the matrix B is bounded by
+        %
+        %                  (1+epsilon)/(1 - epsilon)
+
+        % PROGRAMMING by Carl Christian Kjelgaard Mikkelsen (spock@cs.umu.se)
+        %   2021-10-21  Initial programming and testing.
+
+        % Generate a random matrix
+        R=rand(m,m)-rand(m,m);
+
+        % Eliminate the diagonal
+        D=diag(diag(R)); R=R-D;
+
+        % Measure the weight of the off diagonal entries
+        w=abs(R)*ones(m,1);
+
+        % Construct the new diagonal elements
+        d=w/epsilon;
+
+        % Construct the matrix which is diagonally dominant
+        A=R+diag(d);
+
+        % Do the diagonal scaling
+        B=diag(diag(A))\A;
+
+
+        Once this matrix is generated, we need to alter it a little to make
+        it Hermitian
+    */
+    for (int mat_idx = 0; mat_idx < kMatricesToInvert; mat_idx++) {
```

Comment:
would it make sense to put this into a subroutine to improve readability? This `main()` function is getting a pretty long.